### PR TITLE
partners: VoyageAI rerank

### PR DESCRIPTION
**Pull request by fzowl**
Thank you for contributing to LangChain!

- [x] **PR title**: "package: description"
  - Where "package" is whichever of langchain, community, core, experimental, etc. is being modified. Use "docs: ..." for purely docs changes, "templates: ..." for template changes, "infra: ..." for CI changes.
  - Example: "community: add foobar LLM"

Adding VoyageAI reranking

- [x] **Add tests and docs**: If you're adding a new integration, please include
  1. a test for the integration, preferably unit tests that do not rely on network access,
  2. an example notebook showing its use. It lives in `docs/docs/integrations` directory.

- [x] **Lint and test**: Run `make format`, `make lint` and `make test` from the root of the package(s) you've modified. See contribution guidelines for more: https://python.langchain.com/docs/contributing/

Additional guidelines:
- Make sure optional dependencies are imported within a function.
- Please do not add dependencies to pyproject.toml files (even optional ones) unless they are required for unit tests.
- Most PRs should not touch more than one package.
- Changes should be backwards compatible.
- If you are adding something to community, do not re-import it in langchain.

If no one reviews your PR within a few days, please @-mention one of baskaryan, efriis, eyurtsev, hwchase17.

### CODE REVIEW

**Comment by fzowl** [comment]
@efriis Hello! In my previous PR, I received a comment about enabling contributors(?) changes on the PR. Now i'd like to enable the same :-), but i can't find that option. Can you please help me how i can enable the same?

File: docs/docs/integrations/platforms/voyageai.mdx
Comment (baskaryan) [inline docs/docs/integrations/platforms/voyageai.mdx:1]: could mention the reranker here as well!

File: docs/docs/integrations/retrievers/voyageai-reranker.ipynb
Comment (baskaryan) [inline docs/docs/integrations/retrievers/voyageai-reranker.ipynb:1]: we should put this in integrations/document_transformers (i know we put other rerankers under retrievers, that was an oversight that we'll fix!)

File: libs/partners/voyageai/langchain_voyageai/rerank.py [hunk 1]
```diff
@@ -0,0 +1,144 @@
...
+from voyageai.object import RerankingObject  # type: ignore
+
+
+class VoyageAIRerank(BaseDocumentCompressor):
+    """Document compressor that uses `VoyageAI Rerank API`."""
+
+    client: voyageai.Client = None
+    aclient: voyageai.AsyncClient = None
+    """VoyageAI clients to use for compressing documents."""
+    voyageai_api_key: str
```
Comment (baskaryan) [inline libs/partners/voyageai/langchain_voyageai/rerank.py:21]: should this be a SecretStr?

**Review by baskaryan — COMMENTED** [review]
few comment but overall looks great!

File: docs/docs/integrations/platforms/voyageai.mdx
Comment (sepiatone) [reply docs/docs/integrations/platforms/voyageai.mdx:1]: @baskaryan @fzowl this file is already present in integrations/providers. perhaps just add some info on using the reranker api there?

Comment (fzowl) [reply docs/docs/integrations/platforms/voyageai.mdx:1]: Moved to providers and extended with reranking link.

Comment (baskaryan) [reply docs/docs/integrations/platforms/voyageai.mdx:1]: great catch @sepiatone!

**Review by baskaryan — APPROVED** [review]

**Pull request merged**